nixos/modules/services/web-apps/immich.nix

```
@@ -0,0 +1,297 @@
+        // {
+          default = true;
+        };
+      createDB = mkEnableOption "the automatic creation of the database for immich." // {
+        default = true;
+      };
+      name = mkOption {
+        type = types.str;
+        default = "immich";
+        description = "The name of the immich database.";
+      };
+      host = mkOption {
+        type = types.str;
+        default = "/run/postgresql";
+        example = "127.0.0.1";
+        description = "Hostname or address of the postgresql server. If an absolute path is given here, it will be interpreted as a unix socket path.";
+      };
+      user = mkOption {
+        type = types.str;
+        default = "immich";
+        description = "The database user for immich.";
+      };
+      setupPgvectors = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether to add the pgvecto.rs extension (required to run immich) to postgresql.";
+      };
+    };
+    redis = {
+      enable = mkEnableOption "a redis cache for use with immich" // {
+        default = true;
+      };
+      host = mkOption {
+        type = types.str;
+        default = "/run/immich/redis.sock";
+        description = "The host that redis will listen on.";
+      };
+      port = mkOption {
+        type = types.port;
+        default = 0;
+        description = "The port that redis will listen on. Set to zero to disable TCP.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = !isPostgresUnixSocket -> cfg.secretsFile != null;
+        message = "A secrets file containing at least the database password must be provided when unix sockets are not used.";
+      }
+    ];
+
+    services.postgresql = mkIf cfg.database.enable {
+      enable = true;
+      ensureDatabases = mkIf cfg.database.createDB [ cfg.database.name ];
+      ensureUsers = mkIf cfg.database.createDB [
+        {
+          name = cfg.database.user;
+          ensureDBOwnership = true;
+          ensureClauses.login = true;
+        }
+      ];
+      extraPlugins = mkIf cfg.database.setupPgvectors (ps: with ps; [ pgvecto-rs ]);
+      settings = mkIf cfg.database.setupPgvectors {
+        shared_preload_libraries = [ "vectors.so" ];
+        search_path = "\"$user\", public, vectors";
+      };
+    };
+    systemd.services.postgresql.serviceConfig.ExecStartPost =
+      let
+        sqlFile = pkgs.writeText "immich-pgvectors-setup.sql" ''
+          CREATE EXTENSION IF NOT EXISTS vectors;
+          CREATE EXTENSION IF NOT EXISTS earthdistance CASCADE;
+          ALTER SCHEMA vectors OWNER TO ${cfg.database.user};
+          GRANT SELECT ON TABLE pg_vector_index_stat TO ${cfg.database.user};
+
+          ALTER EXTENSION vectors UPDATE;
+        '';
+      in
+      mkIf cfg.database.setupPgvectors [
+        ''
+          ${config.services.postgresql.package}/bin/psql -d "${cfg.database.name}" -f "${sqlFile}"
```

```suggestion
          ${lib.getExe' config.services.postgresql.package "psql"} -d "${cfg.database.name}" -f "${sqlFile}"
```
Btw, should that be the `mainProgram`?